### PR TITLE
Don't make WebDriverExceptions force-fail the test

### LINKED ISSUE #40057: Webdriver errors always fail the test, even if the test would like to catch them
Currently, because of the `raise` in https://github.com/web-platform-tests/wpt/blob/master/tools/wptrunner/wptrunner/executors/base.py#L749, webdriver errors always lead to a test failure even if the test has a handler for the rejected promise.

I would like to write a test that catches this exception. (my specific use case is polling until a dialog is up, since webdriver does not have events until we get webdriver-bidi eventually)

@gsnedders mentioned on matrix that this would also be helpful for  https://github.com/web-platform-tests/wpt/issues/31782

The concerns mentioned on matrix are that this could mask test failures or lead to timeouts, and may make it harder to debug python-side action failures since we would not print a trace in this situation

Personally I think this is mitigated by:
* test harnesses already have an `unhandledrejection` handler, and if a test has a handler for the rejection it's probably on purpose
* WebDriverException seems unlikely to be caused by a python-side handler and much more likely to come directly from chromedriver/geckodriver/etc, so a python trace would not be helpful

Implementation-wise, it was suggested by @jgraham to have a class variable that's a tuple of exceptions which are non-fatal, similar to the existing `unimplemented_exc`

### PR DESCRIPTION
**Pull request by cbiesinger**
Instead, allow JavaScript to catch them, and have the `unhandledrejection` handler trigger the test failure if they are not caught.

Fixes #40057

### CODE REVIEW

File: tools/wptrunner/wptrunner/executors/base.py [hunk 1]
```diff
@@ -742,8 +743,11 @@ def process_action(self, url, payload):
         except self.unimplemented_exc:
             self.logger.warning("Action %s not implemented" % action)
             self._send_message(cmd_id, "complete", "error", "Action %s not implemented" % action)
+        except self.expected_exc:
+            self.logger.debug("Action %s failed with an expected exception" % action)
+            self._send_message(cmd_id, "complete", "error")
```
Comment (WeizhongX) [inline tools/wptrunner/wptrunner/executors/base.py:748]: Looks like we are still sending out the same message as previous when we get a WebDriverException. So is this change non-functional change besides some log will be different? Does not seem to be the case per the commit message.

Comment (cbiesinger) [reply tools/wptrunner/wptrunner/executors/base.py:748]: The difference is the missing `raise` in this block compared to the final `except` block

Comment (WeizhongX) [reply tools/wptrunner/wptrunner/executors/base.py:748]: Thanks for the explanation.

**Review by WeizhongX — APPROVED** [review]

**Comment by past** [comment]
> I've triggered full WPT runs for Chrome and Firefox in https://wpt.fyi/results/?sha=a3349af546e711927ef32ca1f033549b8358a398, but I don't know how to compare them to runs without this PR?

How about reverting the commit locally and pushing again?

**Comment by cbiesinger** [comment]
As far as I can tell from https://github.com/web-platform-tests/wpt/pull/40058/checks?check_run_id=13875072673 , the full runs on chrome and firefox have passed

**Comment by foolip** [comment]
Given https://wpt.fyi/results/?sha=a3349af546e711927ef32ca1f033549b8358a398, we need to find a run to compare to.

`git merge-base a3349af546e711927ef32ca1f033549b8358a398 origin/master` suggests 54aee0ae47771003af5240de9785199ed10b57bc, and https://wpt.fyi/results/?sha=54aee0ae47771003af5240de9785199ed10b57bc does have results.

From there, turn these URLs into ones with `run_id` using the "copy link" features:

https://wpt.fyi/results/?run_id=5141567151538176&run_id=5071153255677952&run_id=5132997550931968
https://wpt.fyi/results/?run_id=4806242357215232&run_id=5069478486867968&run_id=5143586121711616

Finally, manually shuffle the runs across to get Chrome vs. Chrome and Firefox vs. Firefox:

https://wpt.fyi/results/?run_id=5141567151538176&run_id=4806242357215232
https://wpt.fyi/results/?run_id=5132997550931968&run_id=5069478486867968

In each of those, search for `is:different`.

We suspect regressions would be in WebDriver, so drill down:

https://wpt.fyi/results/webdriver/tests/bidi/script?q=is%3Adifferent&run_id=5141567151538176&run_id=4806242357215232

(There are no WebDriver test differences for Firefox.)

@cbiesinger do any of these changes seem like they could be caused by your changes? @sadym-chromium knows these tests and might be able to help if you need to dig deeper into what the tests are doing.

**Comment by foolip** [comment]
The version of Chrome isn't the same in these two runs, so I suspect that's the real cause, not this change. If someone can confirm, we can merge this PR.

**Comment by cbiesinger** [comment]
@foolip -- thanks for these instructions! Those tests do not go through the "action" interface and so should not be affected. The different Chrome version is likely the reason. Note that if you expand the history, the test was red for several runs.

I think this is ready to merge then! @gsnedders @jgraham ?

**Comment by cbiesinger** [comment]
Also, running the test locally I get the same error with and without my patch.

**Comment by past** [comment]
Thanks for figuring this out! I will merge this on Tuesday if nobody else chimes in by then.

File: tools/wptrunner/wptrunner/executors/executorwebdriver.py [hunk 2]
```diff
@@ -43,6 +43,7 @@
 
 class WebDriverCallbackHandler(CallbackHandler):
     unimplemented_exc = (NotImplementedError, error.UnknownCommandException)
+    expected_exc = (error.WebDriverException,)
```
Comment (jgraham) [inline tools/wptrunner/wptrunner/executors/executorwebdriver.py:46]: We need an equivalent change in `executormarionette.py` (it should be something like `expected_exc = (errors.MarionetteException`)

**Review by jgraham — DISMISSED** [review]

File: tools/wptrunner/wptrunner/executors/executorwebdriver.py (hunk 2 above)
Comment (cbiesinger) [reply tools/wptrunner/wptrunner/executors/executorwebdriver.py:46]: As far as I can tell, marionette does not currently have its own CallbackHander?

Comment (cbiesinger) [reply tools/wptrunner/wptrunner/executors/executorwebdriver.py:46]: Added this.

**Comment by cbiesinger** [comment]
Following foolip's instructions from above, comparing the latest version of this PR (that adds marionette changes) for Firefox:
https://wpt.fyi/results/webdriver?q=is%3Adifferent&run_id=5104707083698176&run_id=5185615912501248

No changes in webdriver/. Some changes in other directories, most likely because the Firefox version has changed.

**Comment by foolip** [comment]
I'll go ahead and merge this. https://github.com/web-platform-tests/wpt/pull/40058#issuecomment-1579311213 shows that it didn't break Firefox.

**Pull request merged**